packages/kbn-router-to-openapispec/src/oas_converter/zod/lib.ts

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { z, isZod } from '@kbn/zod';
+import type { OpenAPIV3 } from 'openapi-types';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import zodToJsonSchema from 'zod-to-json-schema';
+import { KnownParameters } from '../../type';
+import { validatePathParameters } from '../common';
+
+// Adapted from from https://github.com/jlalmes/trpc-openapi/blob/aea45441af785518df35c2bc173ae2ea6271e489/src/utils/zod.ts#L1
+
+const createError = (message: string): Error => {
+  return new Error(`[Zod converter] ${message}`);
+};
+
+function assertInstanceOfZodType(schema: unknown): asserts schema is z.ZodTypeAny {
+  if (!isZod(schema)) {
+    throw createError('Expected schema to be an instance of Zod');
+  }
+}
+
+const instanceofZodTypeKind = <Z extends z.ZodFirstPartyTypeKind>(
+  type: z.ZodTypeAny,
+  zodTypeKind: Z
+): type is InstanceType<typeof z[Z]> => {
+  return type?._def?.typeName === zodTypeKind;
+};
+
+const instanceofZodTypeObject = (type: z.ZodTypeAny): type is z.ZodObject<z.ZodRawShape> => {
+  return instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodObject);
+};
+
+type ZodTypeLikeVoid = z.ZodVoid | z.ZodUndefined | z.ZodNever;
+
+const instanceofZodTypeLikeVoid = (type: z.ZodTypeAny): type is ZodTypeLikeVoid => {
+  return (
+    instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodVoid) ||
+    instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodUndefined) ||
+    instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodNever)
+  );
+};
+
+const unwrapZodType = (type: z.ZodTypeAny, unwrapPreprocess: boolean): z.ZodTypeAny => {
+  if (instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodOptional)) {
+    return unwrapZodType(type.unwrap(), unwrapPreprocess);
+  }
+  if (instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodDefault)) {
+    return unwrapZodType(type.removeDefault(), unwrapPreprocess);
+  }
+  if (instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodLazy)) {
+    return unwrapZodType(type._def.getter(), unwrapPreprocess);
+  }
+  if (instanceofZodTypeKind(type, z.ZodFirstPartyTypeKind.ZodEffects)) {
```

Comment:
Same processing for the 3 scenarios? Seems odd.
Are there any other options that don't require `unwrapZodType`?
Perhaps it'd be interesting to highlight / illustrate them somehow.